Production Docker Image (#195)
nautobot/nautobot: nautobot/docs/docker/index.md

## Getting Started

Nautobot requires a Postgres database and Redis cache before it will start, because of this the quickest and easiest way to get Nautobot running is with docker-compose.  Please see the docker-compose deployment repository for more information.

## Configuration

Most configuration parameters are available via environment variables which can be passed to the container.  If you desire you can inject your own `nautobot_config.py` by overriding `/opt/nautobot/nautobot_config.py`.

### UWSGI

The docker container uses [uWSGI](https://uwsgi-docs.readthedocs.io/) to serve Nautobot.  A default configuration is [provided](/docker/uwsgi.ini), and can be overridden by injecting a new `uwsgi.ini` file at `/opt/nautobot/uwsgi.ini`.  There are a couple of environment variables provided to override some uwsgi defaults:

#### `NAUTOBOT_UWSGI_LISTEN`

Default: `128`

> The socket listen queue size of uWSGI.  In production environments it is recommended to increase this value to 1024 or higher, however depending on your host OS, this may require additional kernel parameter settings, please see [the uWSGI documentation](https://uwsgi-docs.readthedocs.io/en/latest/articles/TheArtOfGracefulReloading.html#the-listen-queue) for more information.

Reviewer: ```suggestion
The socket listen queue size of uWSGI.  In production environments it is recommended to increase this value to 1024 or higher, however depending on your platform, this may require additional kernel parameter settings, please see [the uWSGI documentation](https://uwsgi-docs.readthedocs.io/en/latest/articles/TheArtOfGracefulReloading.html#the-listen-queue) for more information.
```

Please see the [official uWSGI documentation on `listen`](https://uwsgi-docs.readthedocs.io/en/latest/Options.html?highlight=listen#listen) for more information.

---

#### `NAUTOBOT_UWSGI_PROCESSES`

Default: `3`

The number of workers/processes uWSGI will spawn.

Please see the [official uWSGI documentation on `processes`](https://uwsgi-docs.readthedocs.io/en/latest/Options.html?highlight=processes#processes) for more information.

---

### SSL

> Self signed SSL certificates are included by default with the container.  For a production deployment you should utilize your own signed certificates, these can be injected into the container at runtime.  The public certificate should be placed at `/opt/nautobot/nautobot.crt` and the private key should be at `/opt/nautobot/nautobot.key`.

Reviewer: How are certs injected?

> ### Plugins

Reviewer: ```suggestion
### Nautobot Plugins
```

> At this time adding plugins to the existing docker image is not supported, however, you can use this docker image as the base for your Dockerfile to install your own plugins:

Reviewer: ```suggestion
At this time adding Nautobot plugins to the existing Docker image is not supported, however, you can use the included development image as the base within your `Dockerfile` to install your own plugins:
```

## Building the Image

If you have a [development environment](/development/getting-started/#setting-up-your-development-environment) you can use invoke to build the docker images.  By default `invoke build` will build the development containers:

> ```

Reviewer: ```suggestion
```no-highlight
```

$ invoke build
...
$ docker images
REPOSITORY                                                                TAG                    IMAGE ID       CREATED          SIZE
networktocode/nautobot-dev                                                local                  25487d93fc1f   16 seconds ago   630MB
```

If you need to build/test the final image set the `OVERRIDE_FILENAME`:

> ```

Reviewer: ```suggestion
```no-highlight
```

$ export OVERRIDE_FILENAME=docker-compose.build.yml
$ invoke build
...
$ docker images
REPOSITORY                                                                TAG                    IMAGE ID       CREATED          SIZE
networktocode/nautobot                                                    local                  0a24d68da987   55 seconds ago   337MB
```

If you do not have a development environment created you can also build the container using the regular `docker build` command:

> ```

Reviewer: ```suggestion
```no-highlight
```

* `develop-latest-py${PYTHON_VER}` these images are built from the latest code in the develop branch on each commit based on the python version ($PYTHON_VER) docker container
* `develop-${GIT_SHA:0:7}-$(date +%s)-py${PYTHON_VER}` tags for each commit to the develop branch based on the python version ($PYTHON_VER) docker container

To pull a specific tag you can append the image name with `:tag` for example, to pull the 1.0.0 image:

```no-highlight
docker image pull networktocode/nautobot:1.0.0
```

or to pull the 1.0.0 image from GitHub:

```no-highlight
docker image pull ghcr.io/nautobot/nautobot:1.0.0
```

Currently images are pushed for the following python versions:

* 3.6
* 3.7
* 3.8
* 3.9 _For Testing Only_

!!! info
    A dev image `networktocode/nautobot-dev` and `ghcr.io/nautobot/nautobot-dev` is also provided with the same tags, this image provides the development dependencies needed to build Nautobot.  This container can be used as a base for development to develop additional Nautobot plugins but should **NOT** be used in production.

## Getting Started

> Nautobot requires a PostgreSQL database and Redis instance before it will start, because of this the quickest and easiest way to get Nautobot running is with `docker-compose`.  Please see the `docker-compose` deployment repository for more information.

Reviewer 2: What is "the `docker-compose` deployment repository"?

## Configuration

Most configuration parameters are available via environment variables which can be passed to the container.  If you desire you can inject your own `nautobot_config.py` by overriding `/opt/nautobot/nautobot_config.py` using [docker volumes](https://docs.docker.com/storage/volumes/) by adding `-v /local/path/to/custom/nautobot_config.py:/opt/nautobot/nautobot_config.py` to your docker run command, for example:

```no-highlight

> docker run --name nautobot -v /local/path/to/custom/nautobot_config.py:/opt/nautobot/nautobot_config.py ghcr.io/nautobot/nautobot

Reviewer 2: Just above, we recommend using `docker-compose` -- I would have expected this (and the similar SSL example command below) to instruct how to do these things using a docker-compose override instead of a manual Docker command?

Please see the [official uWSGI documentation on `listen`](https://uwsgi-docs.readthedocs.io/en/latest/Options.html?highlight=listen#listen) for more information.

---

#### `NAUTOBOT_UWSGI_PROCESSES`

Default: `3`

The number of worker processes uWSGI will spawn.

Please see the [official uWSGI documentation on `processes`](https://uwsgi-docs.readthedocs.io/en/latest/Options.html?highlight=processes#processes) for more information.

---

### SSL

Self signed SSL certificates are included by default with the container.  For a production deployment you should utilize your own signed certificates, these can be injected into the container at runtime using [docker volumes](https://docs.docker.com/storage/volumes/).  The public certificate should be placed at `/opt/nautobot/nautobot.crt` and the private key should be at `/opt/nautobot/nautobot.key`.  Using a `docker run` these can be injected using the `-v` parameter:

```no-hightlight
docker run --name nautobot -v /local/path/to/custom/nautobot.crt:/opt/nautobot/nautobot.crt -v /local/path/to/custom/nautobot.key:/opt/nautobot/nautobot.key ghcr.io/nautobot/nautobot
```

### Nautobot Plugins

> At this time adding Nautobot plugins to the existing Docker image is not supported, however, you can use the included development image as the base within your `Dockerfile` to install your own plugins:

Reviewer 2: The `:` here suggests there's supposed to be more content here?

# Nautobot Docker Images

Nautobot is packaged as a Docker image for use in a production environment. The published image is based on the `python:3.6-slim` image to maintain the most compatibility with Nautobot deployments. The Docker image and deployment strategies are being actively developed, check back here or join the **#nautobot** Slack channel on [Network to Code](https://networktocode.slack.com) for the most up to date information.

## Tags

The Docker image is published to both the GitHub container registry as well as Docker Hub.

To get the image from Docker Hub run:

```no-highlight

> docker image pull networktocode/nautobot

Reviewer: ```suggestion
$ docker image pull networktocode/nautobot
```

> docker image pull networktocode/nautobot

Author: yes they are the same thing

```

!!! info
    Effective November 20, 2020, [rate limits](https://www.docker.com/increase-rate-limits) have gone into effect on Docker Hub.  To increase the rate limit you can use `docker login` to authenticate with an account, however using the GitHub container registry avoids these limits all together.

To get the image from the GitHub Container Registry (GHCR) run:

```no-highlight

> docker image pull ghcr.io/nautobot/nautobot

Reviewer: ```suggestion
$ docker image pull ghcr.io/nautobot/nautobot
```

```

> 

Author: I'm pretty sure you were the one that was pushing this direction ;) my only $0.02 here is that ghcr.io currently does not have any rate limits while docker hub does.  A lot of organizations are migrating away from docker to either ghcr or quay and therefore in their transition period have containers in both locations.

The following tags are available:

* `X.Y.Z` these images are built with the same baseline as the released Python packages based on the default python version (3.6) docker container
* `latest` these images are built from the latest code in the main branch (should be the latest released version) based on the default python version (3.6) docker container
* `X.Y.Z-py${PYTHON_VER}` these images are built with the same baseline as the released Python packages based on the python version ($PYTHON_VER) docker container
* `latest-py${PYTHON_VER}` these images are built from the latest code in the main branch (should be the latest released version) based on the python version ($PYTHON_VER) docker container
* `develop` these images are built from the latest code in the develop branch on each commit based on the default python version (3.6) docker container
* `develop-${GIT_SHA:0:7}-$(date +%s)` tags for each commit to the develop branch based on the default python version (3.6) docker container

> * `develop-latest-py${PYTHON_VER}` these images are built from the latest code in the develop branch on each commit based on the python version ($PYTHON_VER) docker container

Reviewer 2: In trying to integrate my plugin's CI with these tagged images, I find this one a bit inconvenient:

- Nautobot `1.0.0`, Python `3.6` --> `1.0.0-py3.6`
- Nautobot `1.0.0`, Python `3.7` --> `1.0.0b3-py37`
- Nautobot `develop`, Python `3.8` --> `develop-latest-py38` (have to special-case the `-latest` here)

Would it make sense to change this one to simply `develop-py{PYTHON_VER}`, or is this just something I'll need to special-case in my CI code?

Author: @glennmatthews sorry, someone else pointed this out as well, this doc is out of date `develop-py3.8` is available.

* `develop-${GIT_SHA:0:7}-$(date +%s)-py${PYTHON_VER}` tags for each commit to the develop branch based on the python version ($PYTHON_VER) docker container

To pull a specific tag you can append the image name with `:tag` for example, to pull the 1.0.0 image:

```no-highlight

> docker image pull networktocode/nautobot:1.0.0

Reviewer: ```suggestion
$ docker image pull networktocode/nautobot:1.0.0
```

```

or to pull the 1.0.0 image from GitHub:

```no-highlight

> docker image pull ghcr.io/nautobot/nautobot:1.0.0

Reviewer: ```suggestion
$ docker image pull ghcr.io/nautobot/nautobot:1.0.0
```

```

Currently images are pushed for the following python versions:

* 3.6
* 3.7
* 3.8
* 3.9 _For Testing Only_

!!! info
    A dev image `networktocode/nautobot-dev` and `ghcr.io/nautobot/nautobot-dev` is also provided with the same tags, this image provides the development dependencies needed to build Nautobot.  This container can be used as a base for development to develop additional Nautobot plugins but should **NOT** be used in production.

## Getting Started

> Nautobot requires a PostgreSQL database and Redis instance before it will start, because of this the quickest and easiest way to get Nautobot running is with `docker-compose`.

Reviewer 2: ```suggestion
Nautobot requires a PostgreSQL database and Redis instance before it will start. Because of this the quickest and easiest way to get Nautobot running is with `docker-compose`.
```

> Nautobot requires a PostgreSQL database and Redis instance before it will start, because of this the quickest and easiest way to get Nautobot running is with `docker-compose`.

Reviewer 2: Are we not providing guidelines or an "official" `docker-compose.yml` for production at this time, or did I miss where you've added that?

Reviewer 3: I believe the Docker-Compose for a production like instance is intended to be in a separate repo. Work is commencing on this with this is mind. It is close on a few things, needs some doc TLC, and then update once the container names are available.

Author: Correct, we are specifically leaving out docker-compose from this pr.

## Configuration

Most configuration parameters are available via environment variables which can be passed to the container.  If you desire you can inject your own `nautobot_config.py` by overriding `/opt/nautobot/nautobot_config.py` using [docker volumes](https://docs.docker.com/storage/volumes/) by adding `-v /local/path/to/custom/nautobot_config.py:/opt/nautobot/nautobot_config.py` to your docker run command, for example:

```no-highlight

> docker run --name nautobot -v /local/path/to/custom/nautobot_config.py:/opt/nautobot/nautobot_config.py ghcr.io/nautobot/nautobot

Reviewer: ```suggestion
$ docker run --name nautobot -v /local/path/to/custom/nautobot_config.py:/opt/nautobot/nautobot_config.py ghcr.io/nautobot/nautobot
```

```

Or if you are using docker-compose:

> ```no-highlight

Reviewer: ```suggestion
```yaml
```

Please see the [official uWSGI documentation on `listen`](https://uwsgi-docs.readthedocs.io/en/latest/Options.html?highlight=listen#listen) for more information.

---

#### `NAUTOBOT_UWSGI_PROCESSES`

Default: `3`

The number of worker processes uWSGI will spawn.

Please see the [official uWSGI documentation on `processes`](https://uwsgi-docs.readthedocs.io/en/latest/Options.html?highlight=processes#processes) for more information.

---

### SSL

Self signed SSL certificates are included by default with the container.  For a production deployment you should utilize your own signed certificates, these can be injected into the container at runtime using [docker volumes](https://docs.docker.com/storage/volumes/).  The public certificate should be placed at `/opt/nautobot/nautobot.crt` and the private key should be at `/opt/nautobot/nautobot.key`.  Using a `docker run` these can be injected using the `-v` parameter:

> ```no-hightlight

Reviewer: ```suggestion
```no-highlight
```

> docker run --name nautobot -v /local/path/to/custom/nautobot.crt:/opt/nautobot/nautobot.crt -v /local/path/to/custom/nautobot.key:/opt/nautobot/nautobot.key ghcr.io/nautobot/nautobot

Reviewer: ```suggestion
$ docker run --name nautobot -v /local/path/to/custom/nautobot.crt:/opt/nautobot/nautobot.crt -v /local/path/to/custom/nautobot.key:/opt/nautobot/nautobot.key ghcr.io/nautobot/nautobot
```

```

> Or if you are using docker-compose:

Reviewer: ```suggestion
Or if you are using `docker-compose`:
```

> ```no-highlight

Reviewer: ```suggestion
```yaml
```

services:
  nautobot:
    image: "ghcr.io/nautobot/nautobot"
    volumes:
      - /local/path/to/custom/nautobot.crt:/opt/nautobot/nautobot.crt:ro
      - /local/path/to/custom/nautobot.key:/opt/nautobot/nautobot.key:ro

> ```

Author: If we do support both ghcr and DH I don't think we should provide examples of using both images.  Some of this is basic docker 101

### Nautobot Plugins

At this time adding Nautobot plugins to the existing Docker image is not supported, however, you can use the Nautobot image as the base within your `Dockerfile` to install your own plugins, here is an example dockerfile:

```dockerfile
FROM ghcr.io/nautobot/nautobot

RUN pip install nautobot-chatops

> ```

Reviewer 2: You'd also need to be sure to replace `nautobot_config.py` with one that has the appropriate `PLUGINS` and/or `PLUGIN_CONFIG` values to actually use the plugin.

* `latest` these images are built from the latest code in the main branch (should be the latest released version)
* `latest-dev` these images are built from the latest code in the develop branch on each commit
* `develop-${GIT_SHA:0:7}-$(date +%s)` tags for each commit to the develop branch

## Getting Started

## Configuration

Most configuration parameters are available via environment variables which can be passed to the container.  If you desire you can inject your own `nautobot_config.py` by overriding `/opt/nautobot/nautobot_config.py`.

### Plugins

At this time adding plugins to the existing docker image is not supported, however, you can use this docker image as the base for your Dockerfile to install your own plugins:

## Building the Image

If you have a [development environment](/development/getting-started/#setting-up-your-development-environment) you can use invoke to build the docker images.  By default `invoke build` will build the development containers:

```
$ invoke build
...
$ docker images
REPOSITORY                                                                TAG                    IMAGE ID       CREATED          SIZE

> networktocode/nautobot-dev                                                latest                 25487d93fc1f   16 seconds ago   630MB

Author: switched to local

```

If you need to build/use the production image set the `OVERRIDE_FILENAME`:

```

> $ export OVERRIDE_FILENAME=docker-compose.test.yml

Author: I didn't want to call it `docker-compose.production.yml` as that would be more confusing.  Would "If you need to build/test the final image set the `OVERRIDE_FILENAME`:" be better wording?  I'm open to changing the filename too if you have a better idea

> $ export OVERRIDE_FILENAME=docker-compose.test.yml

Author: swithed to .build.yml

$ invoke build
...
$ docker images
REPOSITORY                                                                TAG                    IMAGE ID       CREATED          SIZE

> networktocode/nautobot                                                    latest                 0a24d68da987   55 seconds ago   337MB

Author: I'm not sure there is any use for this inside of an invoke task, tags are mutable and can be added at will.  I've started adding some examples in the `.travis.yml` there is no need to re-run `invoke build` every time

* 3.9 _For Testing Only_

!!! info

>     A dev image `networktocode/nautobot-dev` and `nautobot/nautobot-dev` is also provided with the same tags, this image provides the development dependencies needed to build Nautobot.  This container can be used as a base for development to develop additional Nautobot plugins but should **NOT** be used in production.

Reviewer: We're not going to be pushing to `nautobot/nautobot` are we? For DH it's `networktocode/natuobot`, yes?

Author: good catch

Most configuration parameters are available via environment variables which can be passed to the container.  If you desire you can inject your own `nautobot_config.py` by overriding `/opt/nautobot/nautobot_config.py` using [docker volumes](https://docs.docker.com/storage/volumes/) by adding `-v /local/path/to/custom/nautobot_config.py:/opt/nautobot/nautobot_config.py` to your docker run command, for example:

```no-highlight

> $ docker run --name nautobot -v /local/path/to/custom/nautobot_config.py:/opt/nautobot/nautobot_config.py nautobot/nautobot

Reviewer: Same here and for the rest of the references to `nautobot/nautobot` as the image.
